mypy/checkexpr.py

```diff
@@ -156,6 +156,11 @@ def visit_name_expr(self, e: NameExpr) -> Type:
     def analyze_ref_expr(self, e: RefExpr, lvalue: bool = False) -> Type:
         result = None  # type: Optional[Type]
         node = e.node
+
+        if isinstance(e, NameExpr) and e.is_special_form:
```

Comment:
Do you know how did this work with the old semantic analyzer (i.e. why do we need this now)?

Reply:
Yes, this is again the story from https://github.com/python/mypy/issues/6458. Old analyzer generates a spurious `Var` as an lvalue node for special form assignments, and thus the type of lvalue is actually inferred (but still `Any` because this is what rvalue returns). The new analyzer doesn't do this, so that we fall through to the end in the `if` just below, thus getting `TypeOfAny.from_error`.